File: controllers/authController.js
const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');
const pool = require('../config/db');

// ========== REGISTER USER OR STAFF ==========
exports.registerUser = async (req, res) => {
  try {
    const {
      first_name,
      last_name,
      email,
      password,
      role,
      roll_number,
      branch
    } = req.body;
    const name = `${first_name} ${last_name}`;

    // 🔍 Debug log to check what role was received from frontend
console.log('Received role from frontend:', role);

    // Find role_id for the provided role
    const [roleRows] = await pool.query('SELECT id FROM roles WHERE name = ?', [role]);
    if (!roleRows.length) return res.status(400).json({ message: 'Invalid role selection.' });

    const role_id = roleRows[0].id;

    // Check if user already exists
    const [userRows] = await pool.query('SELECT * FROM users WHERE email = ?', [email]);
    if (userRows.length) return res.status(409).json({ message: 'Email already registered. Please log in.' });

    // Hash password
    const hashedPassword = await bcrypt.hash(password, 10);

    // Staff require admin approval; students are auto-approved
    let is_approved = true;
    if (role === 'staff') is_approved = false;

    // INSERT user and get inserted user ID
    const [insertResult] = await pool.query(
      `INSERT INTO users
        (name, first_name, last_name, email, password, role_id, is_approved, roll_number, branch)
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
      [name, first_name, last_name, email, hashedPassword, role_id, is_approved, roll_number, branch]
    );

    const newUserId = insertResult.insertId;

    if (role === 'staff') {
      // Staff: do NOT generate token. Wait for approval.
      res.status(201).json({ message: 'Staff registration submitted! Await admin approval.' });
    } else if (role === 'student') {
      // Student: Generate JWT and return user info.
      const token = jwt.sign(
        { id: newUserId, role: role },
        process.env.JWT_SECRET,
        { expiresIn: '1d' }
      );
      // Prepare basic user profile to return
      const userProfile = {
        id: newUserId,
        email,
        role,
        first_name,
        last_name,
        roll_number,
        branch
      };
      res.status(201).json({
        message: 'Registration successful!',
        token,
        user: userProfile
      });
    } else {
      // Unexpected role
      res.status(400).json({ message: 'Invalid role specified.' });
    }
  } catch (err) {
    res.status(500).json({ message: 'Registration failed.', error: err.message });
  }
};


// ========== LOGIN USER, STAFF, OR ADMIN ==========
exports.loginUser = async (req, res) => {
  const { email, password } = req.body;
  try {
    const [rows] = await pool.query(
      'SELECT users.*, roles.name as role FROM users JOIN roles ON users.role_id = roles.id WHERE email = ?',
      [email]
    );
    if (!rows.length) {
      return res.status(404).json({ message: 'No account with this email. Please register.' });
    }

    const user = rows[0];

    // If staff and not approved
    if (user.role === 'staff' && !user.is_approved) {
      return res.status(403).json({ message: 'Staff account is pending admin approval.' });
    }

    const match = await bcrypt.compare(password, user.password);
    if (!match) {
      return res.status(401).json({ message: 'Incorrect password.' });
    }

    // Successful login: generate token and return user info
    const token = jwt.sign(
      { id: user.id, role: user.role },
      process.env.JWT_SECRET,
      { expiresIn: '1d' }
    );

    res.json({
      token,
      user: {
        id: user.id,
        email: user.email,
        role: user.role,
        first_name: user.first_name,
        last_name: user.last_name,
        roll_number: user.roll_number,
        branch: user.branch
        // add more fields here if needed
      }
    });
  } catch (err) {
    res.status(500).json({ message: 'Login failed', error: err.message });
  }
};

File: backend/controllers/authController.py
import os
import bcrypt
import jwt
from flask import request, jsonify
from datetime import datetime, timedelta
from config.db import get_connection 

JWT_SECRET = os.getenv("JWT_SECRET", "your_jwt_secret")

# ========== REGISTER USER OR STAFF ==========
def registerUser():
    conn = None
    cursor = None
    try:
        data = request.get_json()
        first_name = data.get("first_name")
        last_name = data.get("last_name")
        email = data.get("email")
        password = data.get("password")
        role = data.get("role")
        roll_number = data.get("roll_number")
        branch = data.get("branch")

        name = f"{first_name} {last_name}"

        conn = get_connection()
        cursor = conn.cursor(dictionary=True)

        # Find role_id (case-insensitive)
        cursor.execute("SELECT id FROM roles WHERE LOWER(name) = %s", (role.lower(),))
        role_row = cursor.fetchone()
        if not role_row:
            return jsonify({"message": "Invalid role specified."}), 400
        role_id = role_row["id"]

        # Check duplicate email
        cursor.execute("SELECT * FROM users WHERE email = %s", (email,))
        if cursor.fetchone():
            return jsonify({"message": "Email already registered. Please log in."}), 409

        # Hash password
        hashed_password = bcrypt.hashpw(password.encode("utf-8"), bcrypt.gensalt()).decode("utf-8")

        # Staff needs approval, users auto-approved
        is_approved = False if role.lower() == "staff" else True

        # Insert new user
        cursor.execute(
            """
            INSERT INTO users 
                (name, first_name, last_name, email, password, role_id, is_approved, roll_number, branch)
            VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s)
            """,
            (name, first_name, last_name, email, hashed_password, role_id, is_approved, roll_number, branch)
        )
        conn.commit()
        new_user_id = cursor.lastrowid

        if role.lower() == "staff":
            return jsonify({"message": "Staff registration submitted! Await admin approval."}), 201

        elif role.lower() == "user":
            # Generate JWT
            token = jwt.encode(
                {"id": new_user_id, "role": role, "exp": datetime.utcnow() + timedelta(days=1)},
                JWT_SECRET,
                algorithm="HS256"
            )
            user_profile = {
                "id": new_user_id,
                "email": email,
                "role": role,
                "first_name": first_name,
                "last_name": last_name,
                "roll_number": roll_number,
                "branch": branch
            }
            return jsonify({
                "message": "Registration successful!",
                "token": token,
                "user": user_profile
            }), 201

        else:
            return jsonify({"message": "Invalid role specified."}), 400

    except Exception as e:
        return jsonify({"message": "Registration failed.", "error": str(e)}), 500
    finally:
        if cursor:
            cursor.close()
        if conn:
            conn.close()


# ========== LOGIN USER, STAFF, OR ADMIN ==========
def loginUser():
    conn = None
    cursor = None
    try:
        data = request.get_json()
        email = data.get("email")
        password = data.get("password")

        conn = get_connection()
        cursor = conn.cursor(dictionary=True)

        cursor.execute(
            """
            SELECT users.*, roles.name as role 
            FROM users 
            JOIN roles ON users.role_id = roles.id 
            WHERE email = %s
            """,
            (email,)
        )
        user = cursor.fetchone()

        if not user:
            return jsonify({"message": "No account with this email. Please register."}), 404

        # 🚨 Check staff status
        if user["role"].lower() == "staff":
            if user["staff_status"] == "Pending":
                return jsonify({"message": "Staff account is pending admin approval."}), 403
            elif user["staff_status"] == "Rejected":
                return jsonify({"message": "Your staff account was rejected by admin."}), 403

        # Verify password
        if not bcrypt.checkpw(password.encode("utf-8"), user["password"].encode("utf-8")):
            return jsonify({"message": "Incorrect password."}), 401

        # Generate Access & Refresh Tokens
        access_token = jwt.encode(
            {"id": user["id"], "role": user["role"], "exp": datetime.utcnow() + timedelta(minutes=30)},
            JWT_SECRET,
            algorithm="HS256"
        )

        refresh_token = jwt.encode(
            {"id": user["id"], "role": user["role"], "exp": datetime.utcnow() + timedelta(days=7)},
            JWT_SECRET,
            algorithm="HS256"
        )

        return jsonify({
            "access_token": access_token,
            "refresh_token": refresh_token,
            "user": {
                "id": user["id"],
                "email": user["email"],
                "role": user["role"],
                "first_name": user["first_name"],
                "last_name": user["last_name"],
                "roll_number": user["roll_number"],
                "branch": user["branch"],
                "staff_status": user["staff_status"]  # include staff status in response
            }
        }), 200

    except Exception as e:
        return jsonify({"message": "Login failed", "error": str(e)}), 500
    finally:
        if cursor:
            cursor.close()
        if conn:
            conn.close()
